### PR TITLE
CI: Test with negative strides

### PR DESCRIPTION
**Pull request by thedataking**
Add option (`-n`) to test with negative strides similar to dav1d's test configuration. See issue #636 for details.

We test negative strides with the native and optimized x86 and arm64 builds as they take the least time to complete.

### CODE REVIEW

File: .github/workflows/test.sh [hunk 1]
```diff
@@ -63,10 +65,17 @@ else
     test_args+=(--suite testdata_seek-stress)
 fi
 
-if [[ -n $frame_delay ]]; then
+if [[ -n $frame_delay && -n $negative_stride ]]; then
+    # Frame delay is tested with two threads; negative strides with one.
+    echo "Error: frame_delay and negative_stride options can't be used together. Exiting."
```
Comment (kkysen) [inline .github/workflows/test.sh:70]: Can frame delays not be used with negative strides?  Or we're just testing it this one way?

File: src/filmgrain.rs [hunk 2]
```diff
@@ -835,18 +833,16 @@ fn fguv_32x32xn_rust<BD: BitDepth>(
         static W: [[[c_int; 2]; 2 /* off */]; 2 /* sub */] = [[[27, 17], [17, 27]], [[23, 22], [0; 2]]];
 
         let luma_row_uv = |y| {
-            let row = (luma.pixel_offset::<BD>() + luma_row_offset)
+            let row = (luma_row_offset)
```
Comment (kkysen) [inline src/filmgrain.rs:836]:
```suggestion
            let row = luma_row_offset
```

File: src/filmgrain.rs [hunk 3]
```diff
@@ -835,18 +833,16 @@ fn fguv_32x32xn_rust<BD: BitDepth>(
...
                 .wrapping_add_signed((y << sy) as isize * luma.pixel_stride::<BD>());
             luma.slice::<BD, _>((row + (bx << sx).., ..bw << sx))
         };
         let src_row_uv = |y| {
-            let row = (src.pixel_offset::<BD>() + src_row_offset)
-                .wrapping_add_signed(y as isize * src.pixel_stride::<BD>());
+            let row = (src_row_offset).wrapping_add_signed(y as isize * src.pixel_stride::<BD>());
```
Comment (kkysen) [inline src/filmgrain.rs:841]:
```suggestion
            let row = src_row_offset.wrapping_add_signed(y as isize * src.pixel_stride::<BD>());
```

File: src/filmgrain.rs [hunk 4]
```diff
@@ -835,18 +833,16 @@ fn fguv_32x32xn_rust<BD: BitDepth>(
...
             src.slice::<BD, _>((row + bx.., ..bw))
         };
         let dst_row_uv = |y| {
-            let row = (dst.pixel_offset::<BD>() + dst_row_offset)
-                .wrapping_add_signed(y as isize * dst.pixel_stride::<BD>());
+            let row = (dst_row_offset).wrapping_add_signed(y as isize * dst.pixel_stride::<BD>());
```
Comment (kkysen) [inline src/filmgrain.rs:845]:
```suggestion
            let row = dst_row_offset.wrapping_add_signed(y as isize * dst.pixel_stride::<BD>());
```

File: src/filmgrain.rs [hunk 5]
```diff
@@ -708,7 +706,7 @@ fn fgy_32x32xn_rust<BD: BitDepth>(
         };
 
         for y in ystart..bh {
-            let src = &*src_row_y(y);
+            let src: &[<BD as BitDepth>::Pixel] = &*src_row_y(y);
```
Comment (kkysen) [inline src/filmgrain.rs:709]: Why do we need this?

**Review by kkysen — COMMENTED** [review]
Oh whoops about the `fg{y,uv}_32x32xn_rust` bug; I missed that.

For the tests, from your PR description, it sounds like we're not running it on disabled asm builds.  I'm not sure, though.  If we're not, we should.

File: src/filmgrain.rs (hunk 5 above)
Comment (thedataking) [reply src/filmgrain.rs:709]: We don't. I think I accidentally double clicked the type annotation in vscode which inserts the inferred type... sry 😬

File: .github/workflows/test.sh (hunk 1 above)
Comment (thedataking) [reply .github/workflows/test.sh:70]: I don't know enough about how these features interact so so the safest and simplest solution I could come up with was to support the test flows that dav1d has and make sure they don't interfere negatively. We can always come back and do something more sophisticated when there is a clear need.

**Comment by thedataking** [comment]
> For the tests, from your PR description, it sounds like we're not running it on disabled asm builds. I'm not sure, though. If we're not, we should.

We are for the x86 variants. I added a `release-no-asm` variant for macOS/arm64.

**Comment by kkysen** [comment]
> We are for the x86 variants. I added a `release-no-asm` variant for macOS/arm64.

Okay, that sounds good.

**Review by kkysen — APPROVED** [review]
